exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillOptiq.java

```diff
@@ -70,27 +70,65 @@
   private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DrillOptiq.class);
 
   /**
-   * Converts a tree of {@link RexNode} operators into a scalar expression in Drill syntax.
+   * Converts a tree of {@link RexNode} operators into a scalar expression in Drill syntax using one input.
+   *
+   * @param context parse context which contains planner settings
+   * @param input data input
+   * @param expr expression to be converted
+   * @return converted expression
    */
   public static LogicalExpression toDrill(DrillParseContext context, RelNode input, RexNode expr) {
-    final RexToDrill visitor = new RexToDrill(context, input);
+    return toDrill(context, Lists.newArrayList(input), expr);
+  }
+
+  /**
+   * Converts a tree of {@link RexNode} operators into a scalar expression in Drill syntax using multiple inputs.
+   *
+   * @param context parse context which contains planner settings
+   * @param inputs multiple data inputs
+   * @param expr expression to be converted
+   * @return converted expression
+   */
+  public static LogicalExpression toDrill(DrillParseContext context, List<RelNode> inputs, RexNode expr) {
+    final RexToDrill visitor = new RexToDrill(context, inputs);
     return expr.accept(visitor);
   }
 
   private static class RexToDrill extends RexVisitorImpl<LogicalExpression> {
-    private final RelNode input;
+    private final List<RelNode> inputs;
     private final DrillParseContext context;
+    private final List<RelDataTypeField> fieldList;
 
-    RexToDrill(DrillParseContext context, RelNode input) {
+    RexToDrill(DrillParseContext context, List<RelNode> inputs) {
       super(true);
       this.context = context;
-      this.input = input;
+      this.inputs = inputs;
+      this.fieldList = Lists.newArrayList();
+      /*
+         Fields are enumerated by their presence order in input. Details {@link org.apache.calcite.rex.RexInputRef}.
+         Thus we can merge field list from several inputs by adding them into the list in order of appearance.
+         Each field index in the list will match field index in the RexInputRef instance which will allow us
+         to retrieve field from filed list by index in {@link #visitInputRef(RexInputRef)} method. Example:
+
+         Query: select t1.c1, t2.c1. t2.c2 from t1 inner join t2 on t1.c1 between t2.c1 and t2.c2
+
+         Input 1: $0
+         Input 2: $1, $2
+
+         Result: $0, $1, $2
+       */
+      for (RelNode input : inputs) {
```

Comment:
Yes, in `public LogicalExpression visitInputRef(RexInputRef inputRef)` we determine to which input field belongs to. Before that we had only one input thus we did simple get operation `input.getRowType().getFieldList().get(index)` but now we have two inputs so we have to get operation on one input and if field in not found try in the second.  I could iterate over two inputs and do get operation and once filed is found break the loop OR I could merge filed list in one and do simple get operation `fieldList.get(index)`. For performance reasons, I decided to merge filed lists in constructor and use them in `public LogicalExpression visitInputRef(RexInputRef inputRef)` rather than iterating over them for each field.